Move pandas backend to backends/ directory

joy-rosie opened the pull request:
closes #2423 

Discarding previous pull request #2429 as could not figure out a test failing. Most likely a circular import

joy-rosie commented:
I think the tests relating the changes I made are passing, please let me know if there are any specific ones to look at.

Couple things to note:
1) https://github.com/ibis-project/ibis/pull/2440/commits/df38ad80402bc9fd3d8c2e2f3f01bedcf330d68a, I cannot make this import relative as it errors, I think this is due to a circular import but am not 100% sure, let me know if I can try something else
2) In the becnhmarks check: https://github.com/ibis-project/ibis/pull/2440/checks?check_run_id=1197740161#step:9:52, for some reason it is giving the following error:
```
[  0.00%] ···· Traceback (most recent call last):
                 File "/usr/share/miniconda/lib/python3.7/site-packages/asv/benchmark.py", line 860, in disc_modules
                   module = import_module(module_name)
                 File "/home/runner/work/ibis/ibis/.asv/env/9536bcdade9ef9f967582445231c0bf1/lib/python3.7/importlib/__init__.py", line 127, in import_module
                   return _bootstrap._gcd_import(name[level:], package, level)
                 File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
                 File "<frozen importlib._bootstrap>", line 983, in _find_and_load
                 File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
                 File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
                 File "<frozen importlib._bootstrap_external>", line 728, in exec_module
                 File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
                 File "/home/runner/work/ibis/ibis/benchmarks/benchmarks.py", line 6, in <module>
                   from ibis.backends.pandas.udf import udf
               ModuleNotFoundError: No module named 'ibis.backends.pandas'
```
Again I am not sure but this is happening, I tried this locally and seems to work fine.

joy-rosie commented:
I just realised that the benchmark check is failing from it is running from master instead of this pull request

joy-rosie commented:
@datapythonista, just tagging you to the new PR

datapythonista commented on benchmarks/benchmarks.py:
```diff
@@ -156,7 +156,7 @@ def setup(self):
         )
 
-        t = ibis.pandas.connect({'df': self.data}).table('df')
+        t = ibis.backends.pandas.connect({'df': self.data}).table('df')
```
This is because of something very tricky and counter-intuitive in Ibis, but not correct.

If the file imported `ibis.pandas` (instead of `ibis`), the original `ibis.pandas` in this line would be the `ibis.pandas` module. But Since `import ibis` is used, this is the `pandas` attribute in `ibis/__init__.py` (which is the module, but that's irrelevant). That's quite bad practice IMHO, but what it means is that you need to revert this change.

datapythonista commented on docs/source/backends/pandas.rst:
```diff
@@ -4,12 +4,12 @@
 
 Create a client by supplying a dictionary of DataFrames using
-:func:`ibis.pandas.connect`. The keys become the table names:
+:func:`ibis.backends.pandas.connect`. The keys become the table names:
```
Same as before

datapythonista commented on docs/source/backends/pandas.rst:
```diff
@@ -4,12 +4,12 @@
 
    import pandas as pd
-   con = ibis.pandas.connect(
+   con = ibis.backends.pandas.connect(
```
same, please revert

datapythonista commented on ibis/backends/pandas/aggcontext.py:
```diff
@@ -312,7 +312,7 @@ def compute_window_spec_none(_, obj):
 @compute_window_spec.register(dt.Interval)
 def compute_window_spec_interval(_, expr):
-    value = ibis.pandas.execute(expr)
+    value = ibis.backends.pandas.execute(expr)
```
Same as before, can you revert please?

datapythonista commented on ibis/tests/all/test_temporal.py:
```diff
@@ -357,7 +357,7 @@ def test_to_timestamp(backend, con, unit):
 
     ts = ibis.timestamp('2018-04-13 09:54:11.872832')
-    pandas_ts = ibis.pandas.execute(ts).floor(unit).value
+    pandas_ts = ibis.backends.pandas.execute(ts).floor(unit).value
```
Can you revert this too.

datapythonista commented on ibis/tests/backends.py:
```diff
@@ -168,7 +168,7 @@ class Pandas(Backend, RoundHalfToEven):
     @staticmethod
     def connect(data_directory: Path) -> ibis.client.Client:
-        return ibis.pandas.connect(
+        return ibis.backends.pandas.connect(
```
Also this.

datapythonista commented on ibis/tests/expr/test_operations.py:
```diff
@@ -106,7 +106,7 @@ class SpecialTable(ops.DatabaseTable):
             return MyTableExpr
 
-    con = ibis.pandas.connect({})
+    con = ibis.backends.pandas.connect({})
```
And this.

datapythonista reviewed (commented):
Great work @joy-rosie, really nice. There are a couple of couple cases that better to change, but other of that look perfect.

joy-rosie commented:
Sure, I can do that. Although, I have a question related the import. 

In `ibis/__init__.py` when we do `from ibis.backends import pandas`, my understanding is that it allows users to use `ibis.pandas` but when I tried locally, I could only use `ibis.backends.pandas`.

Could you help me understand what `from ibis.backends import pandas` exactly does?

joy-rosie commented:
Just read https://github.com/ibis-project/ibis/pull/2440#discussion_r498865548.

So then my understanding is that if you do `import ibis` then you have access to `ibis.pandas` otherwise you have to use `ibis.backends.pandas`?

And for the user and outside `ibis/backends/pandas` we would prefer to use `ibis.pandas` unless it has explicitly imported `ibis.pandas...`?

joy-rosie commented:
My last question is that why did that not error? If I used `ibis.backends.pandas` from only having run `import ibis` then should it not have errored? Because backends is never imported in `ibis/__init__.py`.

datapythonista commented on ibis/backends/pandas/execution/__init__.py:
```diff
@@ -0,0 +1,11 @@
+from .arrays import *  # noqa: F401,F403
```
Are all these needed for the move to work?

joy-rosie replied:
not necessarily but they were there before so I thought it was worth keeping, I can test if the tests run without them if you like

datapythonista replied:
I didn't see them being moved in the diff, I thought you added them.

It's all right, just wanted to understand both points. Can you merge master and fix conflicts please?

joy-rosie replied:
I merged master, fixed conflicts and pushed but it shows there are still conflicts and the CI is also not running

joy-rosie replied:
sorry, give me a bit of time, I thought I did that but it doesnt seemed to have worked

datapythonista commented on ibis/backends/pandas/dispatch.py:
```diff
@@ -4,11 +4,11 @@
 from multipledispatch import Dispatcher
 
-import ibis
+import ibis.client
```
Why is this change needed?

joy-rosie replied:
this was not needed but thought it would be cleaner this way and a few other files do it this way as well, I can change this back if you like

joy-rosie replied:
Can I resolve this conversation?

datapythonista reviewed (commented):
Looks good, great job @joy-rosie.

Just couple of questions to understand better the changes, but it looks good.

There are couple of extra things that need to be done. We added `isort` to the validation now. Can you run `isort .` before merging master, and then again after merging master (the one before should help git fix the conflicts automatically for you).

Also, we added a `ci/recipe/meta.yaml` file, can you update the `ibis.pandas` path to `ibis.backends.pandas` (you can look for `ibis.backends.omniscidb` for reference).

Thanks!

joy-rosie commented:
> Looks good, great job @joy-rosie.
> 
> Just couple of questions to understand better the changes, but it looks good.
> 
> There are couple of extra things that need to be done. We added `isort` to the validation now. Can you run `isort .` before merging master, and then again after merging master (the one before should help git fix the conflicts automatically for you).
> 
> Also, we added a `ci/recipe/meta.yaml` file, can you update the `ibis.pandas` path to `ibis.backends.pandas` (you can look for `ibis.backends.omniscidb` for reference).
> 
> Thanks!

To clarify, I need to merge from ibis/master into my branch first then run isort?

datapythonista commented:
> To clarify, I need to merge from ibis/master into my branch first then run isort?

I'd do an isort first, I think that would avoid having to fix the conflicts manually. Then you can merge master, and isort again, to make sure the CI is green.

joy-rosie commented:
I think I created a bit of a mess but should be all good now

jreback commented on ibis/backends/pandas/execution/__init__.py:
```diff
@@ -0,0 +1,11 @@
+from .arrays import *  # noqa: F401,F403
+from .decimal import *  # noqa: F401,F403
+from .generic import *  # noqa: F401,F403
+from .join import *  # noqa: F401,F403
```
instead of adding the noqa

add

```
__all__ = ('decimal', 'join', .....)
```

joy-rosie replied:
I tried to do that but got test failures, what exactly do you mean?

I replaced all the lines in that file by __all__ = (...) but that gives failures in the tests

jreback replied:
like this: https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/__init__.py

joy-rosie replied:
I think I did what you have asked for, let me know if that works.

Out of curiosity, why do you want to do that? Is it for speed?

joy-rosie replied:
Sorry I think I tried doing what you suggested then I get the following errors https://github.com/ibis-project/ibis/pull/2440#issuecomment-705865599

jreback replied:
i c, this is not just a list of functions, but much more, we should not do this. 

instead we should figure out what is actually in the public api. ok remove the ``__all__`` and pls create an isseu to followup on this.

joy-rosie replied:
I have made a new issue https://github.com/ibis-project/ibis/issues/2460 and removed `__all__` for now.

pep8speaks commented:
Hello @joy-rosie! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:











* In the file [`ibis/backends/pandas/core.py`](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/core.py):

> [Line 63:80](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/core.py#L63): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 449:80](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/core.py#L449): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)





* In the file [`ibis/backends/pandas/execution/__init__.py`](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py):

> [Line 1:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L1): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .arrays import *' used; unable to detect undefined names
> [Line 2:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L2): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .decimal import *' used; unable to detect undefined names
> [Line 3:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L3): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .generic import *' used; unable to detect undefined names
> [Line 4:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L4): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .join import *' used; unable to detect undefined names
> [Line 5:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L5): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .maps import *' used; unable to detect undefined names
> [Line 6:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L6): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .selection import *' used; unable to detect undefined names
> [Line 7:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L7): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .strings import *' used; unable to detect undefined names
> [Line 8:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L8): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .structs import *' used; unable to detect undefined names
> [Line 9:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L9): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .temporal import *' used; unable to detect undefined names
> [Line 10:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L10): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .timecontext import *' used; unable to detect undefined names
> [Line 11:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L11): [F403](https://duckduckgo.com/?q=pep8%20F403) 'from .window import *' used; unable to detect undefined names
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'join' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'decimal' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'temporal' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'structs' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'window' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'maps' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'timecontext' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'arrays' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'generic' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'strings' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window
> [Line 13:1](https://github.com/ibis-project/ibis/blob/a20d239cd46cc163ab219278d62b809779dfc63f/ibis/backends/pandas/execution/__init__.py#L13): [F405](https://duckduckgo.com/?q=pep8%20F405) 'selection' may be undefined, or defined from star imports: .arrays, .decimal, .generic, .join, .maps, .selection, .strings, .structs, .temporal, .timecontext, .window

jreback commented:
@joy-rosie sorry can you merge master again; we just had a pandas clean up PR go thru. this looks good. ping on green.

joy-rosie commented:
> @joy-rosie sorry can you merge master again; we just had a pandas clean up PR go thru. this looks good. ping on green.

Sure, let me do that.

datapythonista reviewed (approved):
Amazing job @joy-rosie, looks perfect.

jreback commented:
thanks @joy-rosie very nice!
